### PR TITLE
First pass at consolidating types

### PR DESCRIPTION
**Pull request by quinnj**
This will be more nasty than I anticipated, but I've enjoyed digging more into the code. I don't think the package even loads yet, but I've been slowly working through the files. let me know if you see anything already.

### CODE REVIEW

File: src/typedefs.jl [hunk 1]
```diff
@@ -2,105 +2,54 @@ typealias DefaultRefType UInt32
...
-
-            function $P{T, R}(index::Vector{T},
-                              invindex::Dict{T, R},
-                              order::Vector{R})
-                pool = new(index, invindex, order, index[order], V[])
-                buildvalues!(pool)
-                pool
-            end
-        end
+immutable CategoricalPool{T, R <: Integer, V, O}
```
Comment (nalimilan) [inline src/typedefs.jl]: Having "ordered" be a type parameter means it's still necessary to change the type to move from nominal to ordinal. I was thinking of making it a boolean field of `CategoricalPool` and `CategoricalValue` instead.

**Comment by nalimilan** [comment]
Thanks, doesn't look to far from working!

**Comment by nalimilan** [comment]
I've made some progress, tests are passing up to `11_array.jl`. Still need to apply changes to `nullablearray.jl` too. I've pushed my work to `qj/typework` on my fork.

**Comment by nalimilan** [comment]
OK, `nl/typework` now passes the tests.

**Comment by nalimilan** [comment]
I've just merged my branch, please have a look! Thanks for starting this, I thought it would be easier. :-)

Something I'd like to add is a mention "with ordered levels" or something like this when printing the array. Unfortunately, we need to override the default `AbstractArray` printing method for that...

**Comment by quinnj** [comment]
Could we just override the printing of the type and do something like `CategoricalValue{T, R, Ordered=true}` or something?

**Comment by nalimilan** [comment]
Yes, it's just a matter of defining our printing method, but I wonder how much code duplication this entails. Not a big deal, though.

**Pull request closed without merging**